### ISSUE TITLE
[WAGON-537] Maven transfer speed of large artifacts is slow due to unsuitable buffer strategy

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-wagon#591]
**[Olaf Otto](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=o.otto)** opened **[WAGON-537](https://issues.apache.org/jira/browse/WAGON-537?redirect=false)** and commented

We are using maven for build process automation with docker. This sometimes involves uploading and downloading artifacts with a few gigabytes in size. Here, maven's transfer speed is consistently and reproducibly slow. For instance, an artifact with 7,5 GB in size took almost two hours to transfer in spite of a 100 MB/s connection with respective reproducible download speed from the remote nexus artifact repository when using a browser to download. The same is true when uploding such an artifact.

I have investigated the issue using JProfiler. The result shows an issue in AbstractWagon's transfer( Resource resource, InputStream input, OutputStream output, int requestType, long maxSize ) method used for remote artifacts and the same issue in AbstractHttpClientWagon#writeTo(OutputStream).

Here, the input stream is read in a loop using a 4 Kb buffer. Whenever data is received, the received data is pushed to downstream listeners via fireTransferProgress. These listeners (or rather consumers) perform expensive tasks.

Now, the underlying InputStream implementation used in transfer will return calls to read(buffer, offset, length) as soon as **some** data is available. That is, fireTransferProgress may well be invoked with an average number of bytes less than half the buffer capacity (this varies with the underlying network and hardware architecture). Consequently, fireTransferProgress is invoked **millions of times** for large files. As this is a blocking operation, the time spent in fireTransferProgress dominates and drastically slows down the transfers by at least one order of magnitude.

!wagon-issue.png!

In our case, we found download speed reduced from a theoretical optimum of ~80 seconds to to more than 3200 seconds.

From an architectural perspective, I would not want to make the consumers / listeners invoked via fireTransferProgress aware of their potential impact on download speed, but rather refactor the transfer method such that it uses a buffer strategy reducing the the number of fireTransferProgress invocations. This should be done with regard to the expected file size of the transfer, such that fireTransferProgress is invoked often enough but not to frequent.

---

**Affects:** 3.2.0

**Attachments:**
- [wagon-issue.png](https://issues.apache.org/jira/secure/attachment/12943900/wagon-issue.png) (_70.97 kB_)

**Issue Links:**
- [WAGON-543](https://issues.apache.org/jira/browse/WAGON-543) wagon-ssh download hangs
 (_**"causes"**_)
- [MNG-6526](https://issues.apache.org/jira/browse/MNG-6526) Upgrade to Wagon 3.3.1 (from 3.2.0)

**Remote Links:**
- [GitHub Pull Request #50
](https://github.com/apache/maven-wagon/pull/50)
- [GitHub Pull Request #51
](https://github.com/apache/maven-wagon/pull/51)

**Backported to:** [3.3.0](https://github.com/apache/maven-wagon/milestone/33?closed=1)

2 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I will have a look at this by the end of the week. [Olaf Otto](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=o.otto), did you modify the code by disabled all transfer listeners and measuring the performance afterwards? I'd like to know the raw possible performance.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Haven't noticed this yet, but will look into this in the next couple of days. Can you share how this new implementation differs from the previous PR from you?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

What a great improvement! I made some simple tests form my Windows box to my Nexus instance in my LAN with a gigabit link:

```
Before:
Uploaded to nexus-mika: http://mika-ion:8081/nexus/content/repositories/snapshots/test/test-big/0.0.1-SNAPSHOT/test-big-0.0.1-20181110.220953-5-big.bin (5.0 GB at 5.7 MB/s)
Downloaded from nexus-mika: http://mika-ion:8081/nexus/content/repositories/snapshots/test/test-big/0.0.1-SNAPSHOT/test-big-0.0.1-20181110.220953-5-big.bin (5.0 GB at 11 MB/s)

After:
Uploaded to nexus-mika: http://mika-ion:8081/nexus/content/repositories/snapshots/test/test-big/0.0.1-SNAPSHOT/test-big-0.0.1-20181110.215857-3-big.bin (5.0 GB at 20 MB/s)
Downloaded from nexus-mika: http://mika-ion:8081/nexus/content/repositories/snapshots/test/test-big/0.0.1-SNAPSHOT/test-big-0.0.1-20181110.214908-2-big.bin (5.0 GB at 83 MB/s)
```

The difference is insane! I have pushed a slighly modified branch. Any idea why upload is way slower than download?

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

This is Xmas present ... early. Many thanks

**Comment by jira-importer** [comment]
**[Olaf Otto](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=o.otto)** commented

Hi [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)

Great to hear you could reproduce the improvement. I am not sure why the upload does not improve to the same extend in your case - to make sure it's not network related, coud you perhaps compare the upload speed you measured with the upload speed of the respective wget post to your nexus instance? If there is a gap, we'd still have some work to do.

Regards,
Olaf

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I will test the upload with curl in the next couple of days. The [progress monitor](https://github.com/apache/maven/blob/master/maven-embedder/src/main/java/org/apache/maven/cli/transfer/AbstractMavenTransferListener.java) has been reworked by me some time ago. Do you think it makes sense to peg the buffer size to the granularity of the output formatter. Why update the if the decimal does not change?!

**Comment by jira-importer** [comment]
**[Olaf Otto](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=o.otto)** commented

Hi [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)

I agree that it would be great if the progress monitor only made an update if a decimal changes. The only potential issue I see is that if a transfer is slow, users may notice no activity on the console and may believe maven is "hanging".

From an architectural perspective, I would personally not want to create any dependency between TransferListeners and the transfer loop. The loop should simply make sure not to invoke the listeners overly frequently as this slows down transfers (the main objective of this change), and the TransferListeners should do whatever they want with the data.

Perhaps one could discuss whether TransferListeners must be invoked synchronously during data transfer. For me, the term "Listener" implies asynchronism. Are there cases where a listener actively aborts transfer by throwing an exception, perhaps something related to checksums? If not, one could e.g. put progress notifications on a queue and asynchronously notify listeners.

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

sorry about the noise, I am able to cook up a multi SCM freeslyle job to build both wagon and maven

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

Tested with latest maven 3.6.1-SNAPSHOT with both latest wagon snapshot and wagon-2.3.0 currently at staging repo.  Dont see upload improvement. Here one of my 7.1 GB .ova upload  (7.1 GB at 35 MB/s) . Am I missing anything?

Maybe due to my Artifactory and jenkins are 1 network hop away?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran), the upload improvement is not so big as for download. At least with my setup with a low power device hosting Nexus and upload from my Windows machine. How does download look for you? Better? How does curl perform with Artifactory?

I am convinced that [Olaf Otto](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=o.otto)'s work is a great step forward, but not yet complete in terms of performance.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Olaf Otto](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=o.otto), can you have a look at [this](http://mail-archives.apache.org/mod_mbox/maven-dev/201812.mbox/%3CCAPCjjnGSpwPA3%2BaTgy2MHfY7wMLK6_v1ZHBi0RWP-XnjJd5r2w%40mail.gmail.com%3E) message and tell what you think?

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

I see no gain in my download. Basic single download test with curl

curl -O xxxx
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
Dload  Upload   Total   Spent    Left  Speed
100 6779M  100 6779M    0     0   301M      0  0:00:22  0:00:22 --:--:--  343M

maven 3.6.1 with wagon 3.3 is around 115M/s

maven 3.6.0 with wagon 3.2 is around 115M/s

No improvement between 3.6.0/3.6.1.  Did check to make sure wagon-3.3 does landed at my local maven-3.6.1 build

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

How can you see an improvement if curl already fully saturates your gigabit link. How does the download speed look like with Wagon 3.3.0?

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

To test wagon 3.3.0 download speed, I build maven-3.6.1-snapshot with the latest wagon, and run mvn dependency:get to download a single 7G+ file.

Also tested with another artifactory's mirror over a wan link,   the download speed for mvn 3.6, 3.6.1, and curl are the same.  ~60M/s

Note: the network traffic during the holiday is very quiet ( ie no build/test)

**Comment by jira-importer** [comment]
**[Olaf Otto](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=o.otto)** commented

Hi [Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran), thanks a lot for trying this out. So in summary, your tests show no difference between curl, the current maven release version and the version including this change, correct?

If so, you are in a situation where the current implementation already provides the optimal result - this means the buffer in the copy loop is used efficiently and, most importantly, exeuction of fireTransferProgress is **very** fast. So at least we know that the change does not reduce the transfer speed in this situation.

What OS are you on? [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o), same question to you. I've tested Windows 10 so far. I'd assume the costs of e.g. console output might vary greatly.

Regrads,
Olaf

**Comment by jira-importer** [comment]
**[Olaf Otto](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=o.otto)** commented

Hi [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o), regarding your pointer to the mailing list (http://mail-archives.apache.org/mod_mbox/maven-dev/201812.mbox/%3CCAPCjjnGSpwPA3%2BaTgy2MHfY7wMLK6_v1ZHBi0RWP-XnjJd5r2w%40mail.gmail.com%3E), I take it the regression (WAGON-543) is resolved? [Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran) do you still recommend trying to isolate this change to the http implementation?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Olaf Otto](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=o.otto), the regression is actually not in your code, but a bug in JSch.

My tests are on a Windows 10, Intel Skylake against an old 4 thread atom machine running FreeBSD 11.2-RELEASE. I can retry at work with my Windows 7 box against a Xeon based older FreeBSD server. In both cases, server and client are physically very close.

**Comment by jira-importer** [comment]
**[Olaf Otto](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=o.otto)** commented

Thanks [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)!

My hypothesis is that when maven is executed on a *nix derivate (Linux / Mac...), the costs of java.io.Printstream.print(...) to the console can besignificantly lower than on windows, thus this change does not contribute much to the download speed, as it is already almost optimal. However, it the change may still reduce CPU usage due to the reduced number of calls. I'll try and test this on a *nix machine as soon as I have some time.

**Comment by jira-importer** [comment]
**[Olaf Otto](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=o.otto)** commented

I have began testing using docker on win 10 (with hyper-v). For testing, I have run a local docker container via

```java
docker run -i -t -v [local maven installation dir]:/opt/maven -v [local dir with test POM]:/opt/test -v [local .m2 dir]:/root/.m2  openjdk:latest /bin/bash 
```

Then, I have executed a CURL download of a 4GB test file from a remote nexus repo as a reference point. Subsequently, I executed a maven artifact download using maven 3.6.0 with and without the patch. Here are the results:

**Reference download via CURL:** 11 MB/s
**Download with this patch:** 11 MB /s
**Download without patch:** < 1 MB /s

Moreover, the unpatched version caused massive CPU usage due to the millions of invocations of fireTransferProgress. Thus, when using docker for windows, one can see a significant improvement when the remote artifact repo is relatively slow.

I will make another test on a mac to see how this played out on a bare metal *nix.
